### PR TITLE
[tests] update `test_left_padding_compatibility` (and minimize overwrites)

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR updates `test_left_padding_compatibility` so that:
1. We now test left-padding support against all model inputs, as opposed to text-only inputs. This makes the test stricter: we  now test that the model behaves well with left-padded text inputs, together with other (often optional) model inputs. Example: image inputs + padded text inputs may need a corresponding padded cross-attention mask 🔒
2. No true overwrites exist in the library. We can now define custom inputs to the test, to set model-specific inputs that may need to be padded (e.g. see example above) 🧘  

(requirement for #40917 -- the changes made there need the updated test for a happy CI)

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -927,32 +928,44 @@ def test_prompt_lookup_decoding_stops_at_eos(self):
         self.assertTrue(output_prompt_lookup.shape[-1] == 10)
 
     @pytest.mark.generate
-    def test_left_padding_compatibility(self):
-        # NOTE: left-padding results in small numerical differences. This is expected.
-        # See https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535
+    def test_left_padding_compatibility(
```
Comment (gante) [inline tests/generation/test_utils.py:931]: TL;DR
- better documented
- support of custom inputs args (padded and unpadded), for much simpler overwrites
- also prepare padded `token_type_ids`
- forward other model kwargs

File: tests/models/bamba/test_modeling_bamba.py [hunk 2]
```diff
@@ -438,88 +438,11 @@ def test_batching_equivalence(self):
         super().test_batching_equivalence()
         self.model_tester.use_input_mask = orig
 
-    # essentially the same test in test_utils, just adjustment for rtol for this model
```
Comment (gante) [inline tests/models/bamba/test_modeling_bamba.py:441]: this overwrite comment wasn't even true 😢

File: tests/models/idefics/test_modeling_idefics.py [hunk 3]
```diff
@@ -327,7 +326,6 @@ class IdeficsModelTest(ModelTesterMixin, PipelineTesterMixin, GenerationTesterMi
     test_pruning = False
     test_headmasking = False
     test_torchscript = False
-    has_attentions = False  # only supports SDOA and thus no attention probs returned
```
Comment (zucchini-nlp) [inline tests/models/idefics/test_modeling_idefics.py:330]: nit: just a typo 😄

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -963,48 +976,73 @@ def test_left_padding_compatibility(self):
...
             padded_input_ids = torch.cat((padding, input_ids), dim=1)
             padded_attention_mask = torch.cat(
                 (torch.zeros(pad_size[:2], dtype=input_ids.dtype, device=torch_device), attention_mask), dim=1
             )
-            model_kwargs = _prepare_model_kwargs(padded_input_ids, padded_attention_mask, signature)
-            next_logits_with_padding = model(**model_kwargs).logits[:, -1, :]
+            if token_type_ids is not None:
+                padded_token_type_ids = torch.cat(
+                    (
+                        # Assumption: take the first token type id as the padding token type id
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py]: maybe we can use zero as padding token's type. I think that is the most common way for models

File: tests/models/kyutai_speech_to_text/test_modeling_kyutai_speech_to_text.py [hunk 5]
```diff
@@ -361,86 +360,9 @@ def test_disk_offload_safetensors(self):
...
-            padded_attention_mask = torch.cat(
-                (torch.zeros(pad_size[:2], dtype=input_ids.dtype, device=torch_device), attention_mask), dim=1
-            )
-            model_kwargs = _prepare_model_kwargs(padded_input_ids, padded_attention_mask, signature)
-            next_logits_with_padding = model(**model_kwargs).logits[:, -1, :]
-
-            # They should result in very similar logits
-            torch.testing.assert_close(next_logits_wo_padding, next_logits_with_padding, rtol=1e-5, atol=1e-5)
+        # TODO: this tester has non-standard input mokey-patching ☠️
+        unpadded_custom_inputs = self.model_tester.prepare_config_and_inputs_for_common()[1]
```
Comment (zucchini-nlp) [inline tests/models/kyutai_speech_to_text/test_modeling_kyutai_speech_to_text.py]: I didn't really get why we pass through unpadded inputs via overwriting. Maybe we can make comment more clear

**Review by zucchini-nlp — APPROVED** [review]
Great to see that overwritten tests are deleted!

File: tests/models/idefics/test_modeling_idefics.py (hunk 3 above)
Comment (gante) [reply tests/models/idefics/test_modeling_idefics.py:330]: in the deleted code :p

File: tests/models/kyutai_speech_to_text/test_modeling_kyutai_speech_to_text.py (hunk 5 above)
Comment (gante) [reply tests/models/kyutai_speech_to_text/test_modeling_kyutai_speech_to_text.py]: added a better comment

File: tests/models/idefics/test_modeling_idefics.py (hunk 3 above)
Comment (zucchini-nlp) [reply tests/models/idefics/test_modeling_idefics.py:330]: ohh, time to rest from PRs for me 😆

Comment (gante) [reply tests/models/idefics/test_modeling_idefics.py:330]: 😂

File: tests/models/idefics/test_modeling_idefics.py [hunk 6]
```diff
@@ -594,6 +592,33 @@ def test_generate_from_random_inputs_embeds(
...
+            unpadded_custom_inputs=unpadded_custom_inputs, padded_custom_inputs=padded_custom_inputs
+        )
+
+    @unittest.skip(reason="Idefics can't do text-only inference (test filters non-text inputs)")
+    def test_eager_padding_matches_padding_free_with_position_ids(self):
+        pass
+
+    @unittest.skip(reason="Idefics can't do text-only inference (test filters non-text inputs)")
+    def test_sdpa_padding_matches_padding_free_with_position_ids(self):
+        pass
```
Comment (Cyrilvallez) [inline tests/models/idefics/test_modeling_idefics.py:620]: Just a bit confused about those 2 new skips here?

**Review by Cyrilvallez — APPROVED** [review]
Very nice! Let's continue the fight against all those outdated tests and random overwrites! ⚔️
Could you just quickly explain the 2 new skips on idefics?

File: tests/models/idefics/test_modeling_idefics.py (hunk 6 above)
Comment (gante) [reply tests/models/idefics/test_modeling_idefics.py:620]: `has_attentions` was set to `False` on `main`, which skips a few tests. This PR sets it to `True`, which stops skipping tests. This model accepts an attention mask, so it should be set to `True`.

However, these two tests (previously skipped) assume that the model takes `input_ids` as its main inputs, and discard other model inputs for simplicity. IDEFICS needs `pixel_inputs`, and thus the test fails. This test is also skipped in a few other VLMs

Comment (Cyrilvallez) [reply tests/models/idefics/test_modeling_idefics.py:620]: I see, thanks for the explanation!
